my take on https://twistedmatrix.com/trac/ticket/6024

glyph opened the pull request:
https://twistedmatrix.com/trac/ticket/6024

glyph committed:
test utility preparation

add to convenience layer in 'iosim' to allow for deterministic testing
of each step of connection set-up (helpful when testing handshakes...)

glyph committed:
the interface (verbatim from lukasa's branch)

glyph committed:
test support

- pass through Connection mock more cleanly
- add methods that we actually need to call on it during the handshake

glyph committed:
(deterministic!) test case for handshake callback

glyph committed:
implementation, including several simplifying refactorings

glyph committed:
a tale of swords and souls, eternally retold

glyph committed:
move setup into a test utility, test abort-before-send

glyph committed:
eliminate dead code / unnecessary state

glyph committed:
I believe this is correct, but it's unnecessary

(no test can prove its worth)

glyph committed:
move flush to more sensible location

glyph committed:
only do this when I need to do it

glyph committed:
Revert "a tale of swords and souls, eternally retold"

This reverts commit 21fb0e77b350a2e7f3e42d305d3eb047fb0a62a4.

It turns out that this particular error-message distinction _does_
matter; 'tlsv1 unknown ca' is a TLS alert that explains the reason for
the shutdown, whereas 'ssl handshake failure' is just, like, we lost
track of the connection

glyph committed:
the needful

Lukasa commented on twisted/protocols/test/test_tls.py:
```diff
@@ -198,6 +215,104 @@ class NoFactory(object):
+    @ServerFactory.forProtocol
+    @implementer(IHandshakeListener)
+    class Server(Protocol, object):
+        handshaked = False
```
Uh....why is client's version of this called `handshook` and the server's `handshaked`?

glyph replied:
So that they're clearly delineated in the tests and there's no chance that the objects being verified are backwards or duplicated.

Lukasa commented on twisted/protocols/tls.py:
```diff
@@ -321,20 +315,32 @@ def makeConnection(self, transport):
+        try:
+            self._tlsConnection.do_handshake()
+        except WantReadError:
+            self._flushSendBIO()
```
This is maybe just because I know too much about OpenSSL, but it's standing out to me that we don't handle `WantWriteError`. I think I can convince myself that we will never need it, but I'd at least like a comment to float around here that says "We don't need to handle WantWriteError because it should never be emitted."

Lukasa replied:
Alternatively, of course, you can just choose to handle `WantWriteError` with this same block.

glyph replied:
If I'm going to handle `WantWriteError` I'd really want to see a failing test that convinces me that such a thing could ever possibly occur first.  This isn't a new condition though, is it?

glyph replied:
I should note that we don't handle `WantWriteError` _anywhere_ in this file, and as far as I know, we never have.  As I understand it, `WantWriteError` is just OpenSSL's way of saying "`send()` set errno to `EAGAIN`", and since we're using memory BIOs here, OpenSSL isn't allowed to call `send()`.

Lukasa replied:
Yeah: as said, I was able to convince myself that `WantWriteError` can't be raised, but it's the kind of thing that does look weird. Certainly, feel free to leave things unchanged: it's hardly a huge concern. =D

Lukasa commented on twisted/protocols/tls.py:
```diff
@@ -379,23 +385,12 @@ def _flushReceiveBIO(self):
-                if e.args[0] == -1 and e.args[1] == 'Unexpected EOF':
-                    failure = Failure(CONNECTION_LOST)
-                else:
-                    failure = Failure()
```
Why is this gone?

Lukasa replied:
Ah nevermind, I see it got moved.

glyph replied:
It's not gone, just refactored (there's test coverage for this error translation, and if you remove it they fail).  See https://github.com/twisted/twisted/pull/191/files#diff-1f9547fad75f119c2521dbaa75d86ae1R470

glyph committed:
add test case based on lukasa's feedback

This test covers the positive case of the 'if self._appSendBuffer' test
in '_unbufferPendingWrites'; given the early exits in dataReceived for
regular negotiation, this can only be triggered by renegotiation.

glyph committed:
cheating just ever so slightly here

glyph committed:
eliminate some duplicative test methods

glyph committed:
Merge 'origin/trunk' into handshake-6024-2

glyph committed:
assuage pyflakes buildbot
